File: tests/Tests.CSharp/SoftValueCrossVerifyTests.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;

namespace Zeta.Tests.CSharp;

/// <summary>
/// SoftValue cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/soft-value/golden-vectors.json</c>) and asserts identical decisions (resolve /
/// observe-then-resolve). The float confidence/entropy values are not cross-verified (floats don't
/// byte-lock); only the exact decision behavior is.
/// </summary>
public class SoftValueCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(SoftValueCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static Dictionary<string, long> Map(JsonElement obj) =>
        obj.EnumerateObject().ToDictionary(p => p.Name, p => p.Value.GetInt64(), System.StringComparer.Ordinal);

    private static string? Result(JsonElement e) =>
        e.ValueKind == JsonValueKind.Null ? null : e.GetString();

    [Fact]
    public void CSharpSoftValueAgreesWithSeed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "soft-value", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        var seed = doc.RootElement;

        foreach (var v in seed.GetProperty("resolve").EnumerateArray())
        {
            Assert.Equal(
                Result(v.GetProperty("result")),
                SoftValue.Resolve(Map(v.GetProperty("candidates")), v.GetProperty("num").GetInt64(), v.GetProperty("den").GetInt64()));
        }

        foreach (var v in seed.GetProperty("observeResolve").EnumerateArray())
        {
            Assert.Equal(
                Result(v.GetProperty("result")),
                SoftValue.ObserveResolve(
                    Map(v.GetProperty("prior")),
                    Map(v.GetProperty("likelihood")),
                    v.GetProperty("num").GetInt64(),
                    v.GetProperty("den").GetInt64()));
        }
    }
}

File: tests/Tests.FSharp/SoftValue.CrossVerify.Tests.fs
module Zeta.Tests.SoftValueCrossVerifyTests

open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

module SV = Zeta.Core.SoftValue

// ═══════════════════════════════════════════════════════════════════
// SoftValue cross-language agreement — the F# oracle uses the REAL (float) SoftValue and asserts its
// DECISIONS match the shared seed (src/Core.TypeScript/soft-value/golden-vectors.json) that the C#/TS/Rust
// exact-arithmetic oracles also verify. The seed values are chosen far from float-precision boundaries, so
// the decision is float-safe. Only the decisions (resolve / observe-then-resolve) are cross-verified — the
// float confidence/entropy VALUES do not byte-lock and are out of scope.
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private cands (e: JsonElement) : (DynamicValue * float) list =
    [ for p in e.EnumerateObject() -> DynamicValue.String p.Name, float (p.Value.GetInt64()) ]

let private likMap (e: JsonElement) : Map<string, int64> =
    [ for p in e.EnumerateObject() -> p.Name, p.Value.GetInt64() ] |> Map.ofList

let private expected (e: JsonElement) : string option =
    if e.ValueKind = JsonValueKind.Null then None else Some(e.GetString())

let private asString (dv: DynamicValue) : string =
    match dv with
    | DynamicValue.String s -> s
    | _ -> ""

[<Fact>]
let ``F# SoftValue decisions agree with the shared golden seed`` () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "soft-value", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    use doc = JsonDocument.Parse(File.ReadAllText path)
    let root = doc.RootElement
    let section (name: string) : JsonElement[] = root.GetProperty(name).EnumerateArray() |> Seq.toArray

    for v in section "resolve" do
        let threshold = float ((v.GetProperty "num").GetInt64()) / float ((v.GetProperty "den").GetInt64())
        let got =
            SV.ofWeighted (cands (v.GetProperty "candidates"))
            |> Option.bind (SV.resolve threshold)
            |> Option.map asString
        Assert.Equal<string option>(expected (v.GetProperty "result"), got)

    for v in section "observeResolve" do
        let threshold = float ((v.GetProperty "num").GetInt64()) / float ((v.GetProperty "den").GetInt64())
        let lik = likMap (v.GetProperty "likelihood")
        let likelihood (d: DynamicValue) : float =
            match d with
            | DynamicValue.String s -> float (Map.tryFind s lik |> Option.defaultValue 0L)
            | _ -> 0.0
        let got =
            SV.ofWeighted (cands (v.GetProperty "prior"))
            |> Option.bind (SV.observe likelihood)
            |> Option.bind (SV.resolve threshold)
            |> Option.map asString
        Assert.Equal<string option>(expected (v.GetProperty "result"), got)
